Refactor bonded interaction parameter handling

biermanncarl opened the pull request:
Fixes #3019, partial fix for #3680, closes #4066

Description of changes:
- Bonded parameters are now stored inside a ```boost::variant```
- Most bond types are now in self-contained header files (dependency reversals compared to former state)
- Most bond kernels have become members of the respective structs
- Each bond type is represented by its own C++ type (the three tabulated types are now separate, too), which is important because C++ type is now equivalent to bond type
- Removed the former bond type identification scheme from the C++ core
- Separate bonded IA cutoff from non-bonded IA cutoff
- Skip bond loop independently from non-bonded loop

fweik committed:
core: Execute bond loop only when needed

fweik committed:
core: Replace magic number by named constant

KaiSzuttor commented on src/config/config.hpp:
```diff
@@ -90,6 +90,11 @@
 
+/** Maximal number of soft particles per IBM volcons bond (?) */
+#ifndef IBM_MAX_NUM
+#define IBM_MAX_NUM 1000
```
why?

biermanncarl replied:
It's basically a relocation of a constant from [here](https://github.com/espressomd/espresso/pull/4161/files#diff-17cc54295fcd4d75a78d8790cb5c73e2330c7f30d7a504f86ca89c16513bf527L30)

jngrad replied:
This constant had to be moved out to help with dependency inversion, IIRC.

KaiSzuttor commented on src/core/bonded_interactions/angle_cosine.hpp:
```diff
@@ -67,18 +98,16 @@ angle_cosine_force(Utils::Vector3d const &r_mid, Utils::Vector3d const &r_left,
-  auto const k = iaparams.p.angle_cosine.bend;
+  auto const cos_phi0 = this->cos_phi0;
+  auto const sin_phi0 = this->sin_phi0;
+  auto const k = this->bend;
```
these variables are not needed

KaiSzuttor commented on src/core/bonded_interactions/angle_cossquare.hpp:
```diff
@@ -27,32 +27,59 @@
-    auto const k = iaparams.p.angle_cossquare.bend;
+  auto forceFactor = [this](double const cos_phi) {
+    auto const cos_phi0 = this->cos_phi0;
+    auto const k = this->bend;
```
these variables are not needed

KaiSzuttor commented on src/core/bonded_interactions/angle_cossquare.hpp:
```diff
@@ -63,16 +90,14 @@ angle_cossquare_force(Utils::Vector3d const &r_mid,
   auto const cos_phi = std::get<4>(vectors);
-  auto const cos_phi0 = iaparams.p.angle_cossquare.cos_phi0;
-  auto const k = iaparams.p.angle_cossquare.bend;
+  auto const cos_phi0 = this->cos_phi0;
+  auto const k = this->bend;
```
here and elsewhere, please remove those variables, also you do not need to use "this" explicitely, I think we use  the convention of using a "m_" prefix for member variables

biermanncarl replied:
Well, it's a tradeoff. Either we change the name of ALL bonded parameter members, or we keep ```this->```. It's actually only needed in the constructor, and nowhere else if I am not mistaken. I'm unsure what is the better solution.

KaiSzuttor replied:
```this``` is only ever needed for disambiguation. And if you are touching those variables we should change the naming to be consistent with our conventions

jngrad replied:
There is no consensus in core team yet for adding a prefix `m_` to public members. The `this->` dereference however can be removed in all bonded force/energy kernels.

KaiSzuttor commented on src/core/bonded_interactions/bonded_coulomb.hpp:
```diff
@@ -29,51 +29,67 @@
- */
-int bonded_coulomb_set_params(int bond_type, double prefactor);
+/** Parameters for %Coulomb bond Potential */
+struct Bonded_coulomb_bond_parameters {
```
```suggestion
struct BondedCoulombParameters {
```

jngrad replied:
All bonded parameter structs are now CamelCase.

KaiSzuttor commented on src/core/bonded_interactions/bonded_interaction_data.cpp:
```diff
@@ -28,68 +28,29 @@
-                          return std::max(max_cut, cutoff(bond.type, bond.p));
-                        });
+  auto const max_cut_bonded = boost::accumulate(
+      bonded_ia_params, -1.,
```
why do you initialize with -1 and not 0?

jngrad replied:
I think `-1.` is the magic value that will be replaced by the more expressive macro `BONDED_INACTIVE_CUTOFF` in #4066.

KaiSzuttor replied:
Still don't understand this... The minimum value for the maximum cut off is zero and not some magic number.

KaiSzuttor commented on src/core/bonded_interactions/bonded_interaction_data.cpp:
```diff
@@ -99,31 +60,28 @@ double maximal_cutoff_bonded() {
   return (any_dihedrals) ? 2 * max_cut_bonded : max_cut_bonded;
 }
 
+int bonded_ia_params_num_partners(int bond_id) {
```
```suggestion
int no_of_partners(int bond_id) {
```

jngrad replied:
The `bonded_ia_params_` prefix is used in this function and the ones above and below because it's acting on the variant called `bonded_ia_params`. This will become clear once we add the correct doxygen group delimiters. We could even move these functions to `src/script_interface`.

jngrad replied:
Function moved to `src/script_interface/interactions/bonded.hpp` and modified to call a new function with signature `number_of_partners(Bonded_ia_parameters const &iaparams)` in `src/core/bonded_interactions/bonded_interaction_data.cpp`.

jngrad commented on src/core/bonded_interactions/bonded_interaction_data.hpp:
```diff
@@ -23,349 +23,114 @@
-  TabulatedPotential *pot;
+/** These following functions are necessary to hide boost::variant from Cython.
+ *  It can't be used there because Cython doesn't support variadic templates.
+ */
```
It's unclear to which function this doxygen block applies. You can use `@name` and delimiters `/**@{*/` and `/**@}*/` to convey that information ([example](https://github.com/espressomd/espresso/blob/a20d99b94f9ee8c798afdc56f83d51858206beb2/src/core/thermostat.hpp#L58-L67)).

jngrad commented on src/core/bonded_interactions/bonded_interactions.dox:
```diff
@@ -24,131 +24,164 @@
- *     calculation in the integration loop as well as to energy and
- *     pressure tensor analysis
+ *  * C++ core:
+ *    - Create a new header filefor your new bond type, preferably inside the
```
```suggestion
 *    - Create a new header file for your new bond type, preferably inside the
```

jngrad commented on src/core/bonded_interactions/bonded_interactions.dox:
```diff
@@ -24,131 +24,164 @@
- *    @ref Bond_parameters for your interaction defined in the previous step.
- *    For the FENE bond, this would be:
+ *    Replace all the <tt>ar &</tt> commands with the new bond's parameters.
+ *    Every data member of your struct needs to be transmitted. This template
+ *    function can be made private, but then the following line needs to be
+ *    added before for the serialization routine to be able to access it:
```
serialization methods are typically private, therefore I would rephrase it like so:
```suggestion
 *    Every data member of your struct needs to be transmitted. This template
 *    function is private, hence the following line needs to be added before:
```

jngrad commented on src/core/bonded_interactions/bonded_interactions.dox:
```diff
@@ -24,131 +24,164 @@
- *                                       the force calculation and the energy \
- *                                       and pressure analysis
+ *    Refer to @ref fene.hpp as an example.
+
```
```suggestion
```

jngrad commented on src/core/bonded_interactions/bonded_interactions.dox:
```diff
@@ -503,3 +533,58 @@
  *  @f$ \theta_{ijk} = \pi @f$.
  *
  */
+
+/*
```
This break transformed the next block into a regular comment instead of a doxygen comment.
```suggestion
 *
```

jngrad commented on src/python/espressomd/interactions.pxd:
```diff
@@ -298,45 +298,97 @@ IF SOFT_SPHERE:
+            Tabulated_bond_parameters(double min, double max,
+                                      const vector[double] & energy,
+                                      const vector[double] & force) except +
+            int type
```
```suggestion
```

jngrad commented on src/python/espressomd/interactions.pxd:
```diff
@@ -298,45 +298,97 @@ IF SOFT_SPHERE:
+                                      const vector[double] & energy,
+                                      const vector[double] & force) except +
+            int type
+            TabulatedPotential * pot
```
Should this raw pointer become a `shared_ptr[TabulatedPotential] pot` with the appropriate `from libcpp.memory cimport shared_ptr` at the top of the Cython file?

biermanncarl committed:
Refactor Bonded interactions

- store bonded parameters in a boost::variant
- make force/energy kernels methods of the bond structs
- use boost::serialization for MPI communication

Co-authored-by: Jean-Noël Grad <jgrad@icp.uni-stuttgart.de>

jngrad reviewed (dismissed):
good to go from my side

KaiSzuttor commented on src/core/bonded_interactions/angle_cosine.hpp:
```diff
@@ -27,37 +27,65 @@
+  AngleCosineBondParameters(double bend, double phi0);
+
+  std::tuple<Utils::Vector3d, Utils::Vector3d, Utils::Vector3d>
+  angle_force(Utils::Vector3d const &r_mid, Utils::Vector3d const &r_left,
```
i'm not sure about ```angle_force```... wouldn't ```forces``` more precisely reflect that there are multiple forces returned and calculated?

KaiSzuttor commented on src/core/bonded_interactions/angle_cosine.hpp:
```diff
@@ -67,21 +95,17 @@ angle_cosine_force(Utils::Vector3d const &r_mid, Utils::Vector3d const &r_left,
-                                  Utils::Vector3d const &r_right,
-                                  Bonded_ia_parameters const &iaparams) {
+inline double
+AngleCosineBondParameters::angle_energy(Utils::Vector3d const &r_mid,
```
```suggestion
AngleCosineBondParameters::energy(Utils::Vector3d const &r_mid,
```

KaiSzuttor commented on src/core/bonded_interactions/angle_harmonic.hpp:
```diff
@@ -27,36 +27,58 @@
-/** Set parameters for the angle potential. */
-int angle_harmonic_set_params(int bond_type, double bend, double phi0);
+/** Parameters for three-body angular potential (harmonic). */
+struct AngleHarmonicBondParameters {
```
why are these types have a ```Parameters``` suffix if they actually also implement the energy and force kernels?

KaiSzuttor commented on src/core/bonded_interactions/bonded_coulomb_sr.hpp:
```diff
@@ -42,40 +39,58 @@
+  BondedCoulombSRParameters() = default;
+  BondedCoulombSRParameters(double q1q2);
+
+  boost::optional<Utils::Vector3d> pair_force(Utils::Vector3d const &dx) const;
```
```suggestion
  boost::optional<Utils::Vector3d> force(Utils::Vector3d const &dx) const;
```

KaiSzuttor commented on src/core/bonded_interactions/bonded_coulomb_sr.hpp:
```diff
@@ -42,40 +39,58 @@
+  BondedCoulombSRParameters(double q1q2);
+
+  boost::optional<Utils::Vector3d> pair_force(Utils::Vector3d const &dx) const;
+  boost::optional<double> pair_energy(Particle const &p1, Particle const &p2,
```
```suggestion
  boost::optional<double> energy(Particle const &p1, Particle const &p2,
```

KaiSzuttor commented on src/core/bonded_interactions/bonded_interaction_data.hpp:
```diff
@@ -23,346 +23,80 @@
+/** Variant in which to store the parameters of an individual bonded
+ *  interaction
+ */
+using Bonded_ia_parameters = boost::variant<
```
```suggestion
using Bonded_IA_Parameters = boost::variant<
```

KaiSzuttor commented on src/core/bonded_interactions/bonded_tab.cpp:
```diff
@@ -21,66 +21,54 @@
-  if (bond_type < 0)
-    return ES_ERROR;
-
+Tabulated_bond_parameters::Tabulated_bond_parameters(
```
```suggestion
TabulatedBondParameters::TabulatedBondParameters(
```

KaiSzuttor commented on src/core/bonded_interactions/bonded_tab.hpp:
```diff
@@ -32,53 +32,112 @@
-                                double max, std::vector<double> const &energy,
-                                std::vector<double> const &force);
+/** Base class for n-body tabulated potential (n=2,3,4). */
+struct Tabulated_bond_parameters {
```
```suggestion
struct TabulatedBonParameters {
```

KaiSzuttor commented on src/core/bonded_interactions/bonded_tab.hpp:
```diff
@@ -32,53 +32,112 @@
+                                  std::vector<double> const &force);
+  boost::optional<std::tuple<Utils::Vector3d, Utils::Vector3d, Utils::Vector3d,
+                             Utils::Vector3d>>
+  dihedral_force(Utils::Vector3d const &r1, Utils::Vector3d const &r2,
```
```suggestion
  forces(Utils::Vector3d const &r1, Utils::Vector3d const &r2,
```

KaiSzuttor commented on src/core/bonded_interactions/bonded_tab.hpp:
```diff
@@ -32,53 +32,112 @@
-tab_bond_force(Bonded_ia_parameters const &iaparams,
-               Utils::Vector3d const &dx) {
-  auto const *tab_pot = iaparams.p.tab.pot;
+TabulatedDistanceBondParameters::pair_force(Utils::Vector3d const &dx) const {
```
```suggestion
TabulatedDistanceBond::force(Utils::Vector3d const &dx) const {
```

KaiSzuttor reviewed (changes requested):
Please check that all types follow our naming convention (SnakeCase) and adjust the naming of the bond classes (remove the ```Parameters``` suffix)

jngrad committed:
core: Rename bond parameter structs

jngrad committed:
core: Rename bond force/energy kernels

jngrad committed:
Merge branch 'python' into fix-3019

kodiakhq[bot] committed:
Merge branch 'python' into fix-3019
